src/storage/txn/scheduler.rs

```diff
@@ -1076,32 +1233,59 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
         let mut pr = Some(pr);
 
         if !lock_info.is_empty() {
-            assert_eq!(lock_info.len(), 1);
-            let lock_info = lock_info.into_iter().next().unwrap();
-
-            // Only handle lock waiting if `wait_timeout` is set. Otherwise it indicates
-            // that it's a lock-no-wait request and we need to report error
-            // immediately.
-            if lock_info.parameters.wait_timeout.is_some() {
-                assert_eq!(to_be_write.size(), 0);
-                pr = Some(ProcessResult::Res);
-                // allow_lock_with_conflict is not supported yet in this version.
-                assert!(!lock_info.parameters.allow_lock_with_conflict);
-
-                scheduler.on_wait_for_lock(&ctx, cid, lock_info, tracker);
+            if tag == CommandKind::acquire_pessimistic_lock {
+                assert_eq!(lock_info.len(), 1);
+                let lock_info = lock_info.into_iter().next().unwrap();
+
+                // Only handle lock waiting if `wait_timeout` is set. Otherwise it indicates
+                // that it's a lock-no-wait request and we need to report error
+                // immediately.
+                if lock_info.parameters.wait_timeout.is_some() {
+                    assert_eq!(to_be_write.size(), 0);
+                    pr = Some(ProcessResult::Res);
+
+                    scheduler.on_wait_for_lock(&ctx, cid, lock_info, tracker);
+                } else {
+                    // For requests with `allow_lock_with_conflict`, key errors are set key-wise.
+                    // TODO: It's better to return this error from
+                    // `commands::AcquirePessimisticLocks::process_write`.
```

Comment:
So.., if we are setting locks in `pr` here now,  what's the meaning of setting a `PessimisticLockKeyResult::Failed` to `pr` in acquire_pessimistic_locks?

Reply:
Though we support only single-key requests for now, the protocol is at first designed to support processing multiple keys in one request. In multiple keys cases, it's unavoidable that we may fail to lock some of the keys while the other keys is successfully locked. To allow patially failing, this type is added so that we can see in the resp.results that which key is failed.

The remaining missing parts are:

* deadlock detection
* handling the case that timeout when the request is resumed and being executing, but still meet lock (in current implementation, waiter manager will simply drop the waiter since the entry can't be found in the lock waiting queue; if it's added to the queue again, it will then never timeout).

Reply:
Processing a single key is already difficult enough for us.. :sweat_smile: